### DISCUSSION TITLE
Icons are misaligned with a pitched Maplibre base map

### DISCUSSION THREAD

**Discussion by hodossy** [visgl/deck.gl#7054] [Q&A] [answered]
I have the following interleaved setup in an Angular app:

```console
npm i @deck.gl/core @deck.gl/extensions @deck.gl/layers @deck.gl/mapbox
```

```javascript
const map = new Map({
  container: hostElement,
  style: "style-url-here",
  interactive: false,
  center,
  bearing,
  zoom,
  pitch: 60,
  fadeDuration: 0,
});

const deck = new Deck({
  gl: map.painter.context.gl,
  onClick: (info: any) => onEntityClicked(info.object),
  getCursor: (state: any) => (state.isHovering ? 'pointer' : 'grab'),
  parameters: {
    depthTest: false,
  },
});

const myDeckLayer = new MapboxLayer({ id: 'deckgl', deck: this.deck });
map.addLayer(myDeckLayer);
```

The whole app is simulating a car drive, in a pitched view, where the base map is rotated according to current heading. The 

I have seen [this issue](https://github.com/visgl/deck.gl/issues/3247), but I don't use the viewState property, nor I have any altitude parameters in my data, just lon/lat. I would greatly appretiate any suggestions where the issue might be, I am quite new to deck.gl.

**Comment by Pessimistress** [comment, marked as answer]
Do you have the same problem without using MapboxLayer? If so, can you create a codepen?

**Reply by hodossy** [reply]
Thanks for the tip, I have rewritten my app to use an overlayed deck, and I have got rid of the MapboxLayer class. Unfortunately the situation did not improved, but I managed to resolve another unrelated issue with this! I am working on the codepen repro, but I need to create some   mock data an so on. In the meanwhile, I am using this code to update the viewstate (update the center, the zoom and bearing, while pitch is the same)

```typescript
  const center = map.getCenter();
  this.deck = new Deck({
    canvas: 'deck-canvas',
    width: '100%',
    height: '100%',
    controller: false,
    onClick: (info: any) => onEntityClicked(info.object),
    getCursor: (state: any) => (state.isHovering ? 'pointer' : 'grab'),
    parameters: {
      depthTest: false,
    },
    initialViewState: {
      latitude: center.lat,
      longitude: center.lng,
      zoom: map.getZoom(),
      bearing: map.getBearing(),
      pitch: map.getPitch(),
    },
    onViewStateChange: ({ viewState }: any) => {
      map.jumpTo({
        center: [viewState.longitude, viewState.latitude],
        zoom: viewState.zoom,
        bearing: viewState.bearing,
        pitch: viewState.pitch,
      });
    },
  });

  setViewState(viewState: ViewState) {
    this.deck.setProps({ viewState, transitionDuration: 100 }); // I needed to set a transition, otherwise nothing happened
    this.deck.props.onViewStateChange({ viewState }); // if I omitted this call, the base map has not been updated, since the callback is never called
  }
```

**Reply by hodossy** [reply]
I have created a minimal 'repro' [on CodePen](https://codepen.io/hodossy/pen/ExQzKzx), but the issue does not show. I tried to do the same with a minimal Angular app, but the issue still does not show, so I definitely have an implementation issue, I just cannot point it out yet.

**Reply by hodossy** [reply]
Now I managed to find the issue, the problem is that I use a `200px` [padding on the map](https://maplibre.org/maplibre-gl-js-docs/api/map/#map#setpadding), like `map.setPadding({ top: 200, bottom: 0, left: 0, right: 0 })`. Is there a similar option for deck.gl?

**Reply by hodossy** [reply]
I have also updated the CodePen to show the issue, and I have found #5359 relating to this one.

**Reply by hodossy** [reply]
I tried to add `projectionMatrix` parameter to the `viewState`, but no success. Could you please show me the direction to implement my custom view? I suppose I need to subclass [WebMercatorViewport](https://deck.gl/docs/api-reference/core/web-mercator-viewport), and add the required translation to the projection matrix, [similar to this code snippet](https://github.com/visgl/deck.gl/blob/8.7-release/modules/core/src/viewports/web-mercator-viewport.js#L118):

```javascript
if (worldOffset) {
  const viewOffset = new Matrix4().translate([512 * worldOffset, 0, 0]);
  viewMatrixUncentered = viewOffset.multiplyLeft(viewMatrixUncentered);
}
```

I am just not sure how to pass my custom viewport to the deck instance, and what unit the `translate` parameters are in. Why is that 512 there? Many thanks!

**Reply by Pessimistress** [reply]
Padding is supported in the latest beta (`^8.8.0-beta`). You can pass the following to `Deck`:

```js
views: new MapView({padding: { top: 200, bottom: 0, left: 0, right: 0 }})
```

https://github.com/visgl/deck.gl/pull/6672/files

**Reply by hodossy** [reply]
Many thanks, then I just wait for the release! Thanks!
